File: 3956-maximum-sum-of-m-non-overlapping-subarrays-i/3956-maximum-sum-of-m-non-overlapping-subarrays-i.scala
object Solution {
    def maximumSum(nums: Array[Int], m: Int, l: Int, r: Int): Long = {
        val n = nums.length

        val pre = Array.fill[Long](n + 1)(0L)
        for (i <- 0 until n) {
            pre(i + 1) = pre(i) + nums(i)
        }

        var cur = Array.fill[Long](n + 1)(0L)
        var ans = Long.MinValue

        for (_ <- 1 to m) {
            val nxt = Array.fill[Long](n + 1)(Long.MinValue)

            val dq = new Array[Int](n + 1)
            var h = 0
            var t = -1

            for (i <- l to n) {
                val idx = i - l

                if (cur(idx) != Long.MinValue) {
                    while (h <= t &&
                        cur(dq(t)) - pre(dq(t)) <= cur(idx) - pre(idx)) {
                        t -= 1
                    }
                    t += 1
                    dq(t) = idx
                }

                while (h <= t && dq(h) < i - r) {
                    h += 1
                }

                if (h <= t) {
                    nxt(i) = cur(dq(h)) + pre(i) - pre(dq(h))
                    if (nxt(i) > ans) ans = nxt(i)
                }
            }

            var mx = Long.MinValue
            val tmp = Array.fill[Long](n + 1)(Long.MinValue)

            for (i <- 0 to n) {
                mx = math.max(mx, nxt(i))
                tmp(i) = mx
            }

            cur = tmp
        }

        ans
    }
}

File: 3956-maximum-sum-of-m-non-overlapping-subarrays-i/3956-maximum-sum-of-m-non-overlapping-subarrays-i.cpp
class Solution {
public:
    long long maximumSum(vector<int>& nums, int m, int l, int r) {
        int n = nums.size();

        vector<long long> pre(n + 1);
        for (int i = 0; i < n; i++) pre[i + 1] = pre[i] + nums[i];

        const long long INF = -(long long)4e18;

        vector<vector<long long>> dp(m + 1, vector<long long>(n + 1, INF));

        for (int i = 0; i <= n; i++) dp[0][i] = 0;

        long long ans = INF;

        for (int k = 1; k <= m; k++) {
            deque<pair<int,long long>> dq;

            for (int j = 1; j <= n; j++) {

                int p = j - l;

                if (p >= 0 && dp[k - 1][p] != INF) {
                    long long v = dp[k - 1][p] - pre[p];

                    while (!dq.empty() && dq.back().second <= v)
                        dq.pop_back();

                    dq.push_back({p, v});
                }

                while (!dq.empty() && dq.front().first < j - r)
                    dq.pop_front();

                dp[k][j] = dp[k][j - 1];

                if (!dq.empty())
                    dp[k][j] = max(dp[k][j], dq.front().second + pre[j]);
            }

            ans = max(ans, dp[k][n]);
        }

        return ans;
    }
};